### PR TITLE
DOC: add quickstart section to the gridspec tutorial

### PR DESCRIPTION
**Pull request by phobson**
## PR Summary

This PR adds basic example showing off the gridspec module. 
Here's a notebook of the plots: https://gist.github.com/phobson/b087eabac6eda78986bc643e7aaea4ae

## PR Checklist

- [x] Code is PEP 8 compliant 
- [x] Documentation is sphinx and numpydoc compliant

### CODE REVIEW

**Comment by Tillsten** [comment]
I think one can find gridspec now in the pyplot namespace.

**Comment by tacaswell** [comment]
This should include a link to http://matplotlib.org/users/gridspec.html?highlight=gridspec

**Comment by phobson** [comment]
@tacaswell Thanks for the link. I couldn't find that last night. Do you know off the top of your head what relative path is at build time? (I'm really bad at sphinx). I can check later today when I'm at home on my mac.

@Tillsten Thanks for the note. I didn't know that. The tutorial Tom linked to uses `import matplotlib.gridspec as gridspec`, so I'll stay consistent with that.

File: examples/subplots_axes_and_figures/basic_gridspecs.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+"""
+===================================================
+Advanced subplot placement with the gridspec module
+===================================================
+
+The ``subplots`` function is good for building a uniform grid
+of axes. The ``gridspec`` module provides more control, but is
+proportionally more complex to use.
+
+The API documentation for the ``gridspec`` is `here <users/gridspec>`_.
```
Comment (phobson) [inline examples/subplots_axes_and_figures/basic_gridspecs.py]: perhaps `../../users/gridspec`

**Comment by choldgraf** [comment]
@phobson this is cool! good use of SG. since the SG PR is about to be merged, could you modify the `gridspec` api page so that it lists the actual classes in the module w/ `autosummary` (see the `patches` api page for example). This will cause the API page to auto-link to the places where it's used in the examples!

**Comment by choldgraf** [comment]
@phobson actually you know what, this won't work until the other PR is actually merged so I'll just change it there myself to avoid confusion :)

**Comment by choldgraf** [comment]
OK all the SG PRs are merged now. One quick thought: this seems more like a "tutorial" than an example. What about putting this in [tutorials](http://matplotlib.org/devdocs/tutorials/index.html) in the intermediate section? At some point it might even be worth spinning off this + the `tight_layout` tutorial into their own folder called "Controlling Figure / Plot Layouts" or something.

**Comment by phobson** [comment]
@choldgraf I think that's a great idea. I'll rebase and update hopefully later today

**Comment by story645** [comment]
How is this different though then from the [subplot/gridspec](http://predictablynoisy.com/matplotlib/tutorials/02_intermediate/gridspec.html#sphx-glr-tutorials-02-intermediate-gridspec-py) tutorial that's already in there? Or can it be incorporated into it?

File: examples/subplots_axes_and_figures/basic_gridspecs.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+`here <../../users/gridspec.html>`_.
+
+"""
+
+
+import matplotlib.pyplot as plt
+import matplotlib.gridspec as gridspec
+
+
+SIZE = (5, 5)
```
Comment (anntzer) [inline examples/subplots_axes_and_figures/basic_gridspecs.py]: Do you need to pass SIZE explicitly?  it seems irrelevant to this discussion.

File: examples/subplots_axes_and_figures/basic_gridspecs.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
+"""
+===================================================
+Advanced subplot placement with the gridspec module
+===================================================
+
+The ``subplots`` function is good for building a uniform grid
```
Comment (anntzer) [inline examples/subplots_axes_and_figures/basic_gridspecs.py]: Should use single backquotes here and below and link to the actual function/module.  The additional sentence regarding the API doc would then become unneeded.

**Comment by choldgraf** [comment]
@story645 I think this is a good point - there's some information that is overlapping and others that isn't. I like the narrative style of this one a bit more, I think it's easier to follow. 

I guess one question would be: is there a natural place to split up this tutorial + the other one so that we have 2 tutorials, or alternatively is there a way to incorporate this tutorial into the material of the other? I like @phobson 's style of writing so I am partial to having the other tutorial be improved, but I also don't want to volunteer him for more work ;-)

**Comment by phobson** [comment]
@choldgraf perhaps it makes sense to trim this down a bit, and make it a "gridspec quickstart" section of the existing tutorial, along with going over that one's existing content

**Comment by anntzer** [comment]
Two remarks regarding the already existing gridspec tutorial:
- there are some markup issues.
- calling gs.update(...) is not needed when you can directly pass the parameters to the constructor.

**Comment by choldgraf** [comment]
@phobson I think that's a great idea - that tutorial could definitely be made better, and I think that if the structure / content is improved and made more clear, it'll be more effective in the long run than adding a new tutorial. Happy to review etc if you give it a shot.

**Comment by choldgraf** [comment]
@phobson btw, I found that it can be useful to show off how new content looks on the built docs by using `ghp-import` to push to `gh-pages` on your matplotlib fork, and then hosting that with github so you can send links etc

**Comment by phobson** [comment]
@choldgraf @story645 I took a first pass at merging my example with the existing gridspec tutorial. if you get a sec, please let me know what you think about the direction I'm taking this. 

My next steps would be to add a little more explanatory text around the existing examples and unify the coding styles a bit.

File: tutorials/02_intermediate/gridspec.py [hunk 4]
```diff
@@ -1,59 +1,162 @@
...
+fig1.tight_layout()
+
+############################################################################
+# For such a simple use case, :mod:`~matplotlib.gridspec` is perhaps overly
+# verbose.
+# You have to create the figure and :class:`~matplotlib.gridspec.GridSpec`
+# instance separately, then pass the elements gridspec instance to the
+# :func:`~matplotlib.figure.Figure.add_subplot` method to create the axes
+# objects.
+# The elements of the gridspec are accessed as if they where elements of
```
Comment (story645) [inline tutorials/02_intermediate/gridspec.py]: type: were, not where

File: tutorials/02_intermediate/gridspec.py [hunk 5]
```diff
@@ -1,59 +1,162 @@
...
+# When you want to have plots of different sizes, however,
+# :mod:`~matplotlib.gridspec` becomes indispensable and provides a couple
+# of options.
+# The method shown here initializes a uniform grid specification,
+# and then uses typical numpy indexing and slices to allocate multiple
+# "cells" for a given subplot.
+
+fig3 = plt.figure()
+spec3 = gridspec.GridSpec(ncols=3, nrows=3)
+anno_opts = dict(xy=(0.5, 0.5), xycoords='axes fraction',
```
Comment (story645) [inline tutorials/02_intermediate/gridspec.py]: an explanation, could just be conceptual, of these keywords might be helpful

File: tutorials/02_intermediate/gridspec.py [hunk 6]
```diff
@@ -1,126 +1,196 @@
...
+# :class:`~matplotlib.axes.Axes` objects.
+
+fig1, f1_axes = plt.subplots(ncols=2, nrows=2)
+fig1.tight_layout()
+
+############################################################################
+# For such a simple use case, :mod:`~matplotlib.gridspec` is perhaps overly
+# verbose.
+# You have to create the figure and :class:`~matplotlib.gridspec.GridSpec`
+# instance separately, then pass the elements gridspec instance to the
```
Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: it seems like you've a word

"elements of the" maybe?

File: tutorials/02_intermediate/gridspec.py (hunk 5 above)
Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: and maybe just add a comment to remind people that `add_subplot` returns an `axes` object so you're chaining them together with `annotate`

File: tutorials/02_intermediate/gridspec.py [hunk 7]
```diff
@@ -1,126 +1,196 @@
...
+
+fig3.tight_layout()
+
+############################################################################
+# This example shows off using the ``width_ratios`` and ``height_ratios``
+# options.
+# These keyword arguments are lists of numbers.
+# Note that absolute values are meaningless, only their relative ratios
+# matter.
+# That means that ``width_ratios=[2, 4, 8]`` is equivalent to
```
Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: I'm not sure, but will this actually render as a new line in the rST? I think we need an explicit space

File: tutorials/02_intermediate/gridspec.py [hunk 8]
```diff
@@ -1,126 +1,196 @@
...
+# useful since :func:`~matplotlib.pyplot.subplots` accepts them within the
+# ``gridspec_kw`` parameter.
+# For that matter, any parameter accepted by
+# :class:`~matplotlib.gridspec.GridSpec` can be passed to
+# :func:`~matplotlib.pyplot.subplots` via the ``gridspec_kw`` parameter.
+# This example recreates the previous figure without directly using a
+# gridspec instance.
+
+gs_kw = dict(width_ratios=widths, height_ratios=heights)
+fig5, f5_axes = plt.subplots(ncols=3, nrows=3, gridspec_kw=gs_kw)
```
Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: this is magic, I didn't know you could do this :-P

File: tutorials/02_intermediate/gridspec.py [hunk 9]
```diff
@@ -1,126 +1,196 @@
...
+        ax.annotate(label, (0.1, 0.5), xycoords='axes fraction', va='center')
+
+fig5.tight_layout()
 
+
+###############################################################################
+# Using subplot2grid
+# ==================
+#
 # To use :func:`~matplotlib.pyplot.subplot2grid`, you provide geometry of
```
Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: maybe one conceptual-level sentence about what `subplot2grid` is relative to gridspec, and why it exists

File: tutorials/02_intermediate/gridspec.py [hunk 10]
```diff
@@ -1,126 +1,196 @@
...
-ax6 = plt.subplot(gs2[-1, -1])
+
+gs2 = gridspec.GridSpec(nrows=3, ncols=3, left=0.55, right=0.98,
+                        hspace=0.05, fig=fig)
+ax4 = fig.add_subplot(gs2[:, :-1])
+ax5 = fig.add_subplot(gs2[:-1, -1])
+ax6 = fig.add_subplot(gs2[-1, -1])
 
 ###############################################################################
 # GridSpec using SubplotSpec
```
Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: here it's also be good to see some conceptual-level sentence or two about what subplotspec is, why it exists / how it's different from gridspec, etc

**Comment by choldgraf** [comment]
really like these changes @phobson ! I added a few comments.

One general thought I had is that I feel like these tutorials should go from least -> most complicated as you read through. To that extent do you think it'd be worth moving `subplot2grid` earlier on, since (at least in my mind) it's conceptually a little simpler than using gridspec? I'm not sure if this is the best way to do it but just a thought.

File: tutorials/02_intermediate/gridspec.py (hunk 7 above)
Comment (phobson) [inline tutorials/02_intermediate/gridspec.py]: I don't want a new line (and there is not one in my build). When writing in plaintext, markdown, RST, whatever, I like to put each sentence on a its own line (long senterences might be hard-wrapped or not). So when you make changes, you only have to reflow the sentence, not the paragraph, and don't see any noisey, irrelevant changes to other sentences in the diff.

Comment (choldgraf) [inline tutorials/02_intermediate/gridspec.py]: Cool - that makes sense to me! Just noting it to make sure that was intended behavior! (btw, feel free to delete comments of mine that you address or aren't actionable...I tend to do this as a way of ticking off "todo" items in a PR)

File: tutorials/02_intermediate/gridspec.py (hunk 9 above)
Comment (jklymak) [reply tutorials/02_intermediate/gridspec.py]: As someone who doesn't think subplot2grid should exist, I'd advocate moving this section down or removing it altogether.  `subplot2grid` uses `gridspec`, but only in a somewhat peverted way, using a new `gridspec` for each call.  That loses any of the logical layout gridspec gives you, so its a strange thing to highlight early in the gridspec tutorial.

**Comment by choldgraf** [comment]
re: my earlier comment on `subplot2grid`, I think it's more important to get this merged so I'm happy to go w/ @jklymak 's suggestion

**Comment by jklymak** [comment]
I wasn’t trying to veto anything with that comment about subplot2grid.  This or should go in regardless.  I just think it’s a bit orthogonal to gridspec.

**Comment by choldgraf** [comment]
ping @phobson - travis is unhappy FYI. Other than that, is this ready to go from your end?

**Comment by phobson** [comment]
I'll rebase this tonight and shift the sections around based on the rest of
the feedback (which is much appreciated!)
-paul

**Comment by choldgraf** [comment]
ping :-) lemme know when the rebase is there

**Comment by jklymak** [comment]
@phobson popping back onto your queue, as I think this is a great change (but needs a rebase!)

**Comment by choldgraf** [comment]
@phobson ping us here when you're ready for another look

**Comment by phobson** [comment]
@choldgraf @jklymak rebased and made some edits.

Off for a bike ride. Back in 6 hours 😉

**Comment by phobson** [comment]
(i.e., ready for another look)

**Comment by choldgraf** [comment]
As we already took a look at this before I think we should just merge once the tests are passing. Any idea what's up with circle?

**Comment by jklymak** [comment]
...bad indent in `gridspec.py`, but hard to debug w/o building locally.

**Comment by phobson** [comment]
I fixed(?) the whitespace error in gridspec.py that was propagating to gridspec.rst, but now the build errors are nonsense to me.

**Comment by choldgraf** [comment]
This feels like something that's not related to the PR...any idea what's going on @tacaswell ? Did the travis infrastructure change recently?

**Comment by jklymak** [comment]
Hard-cycling

**Comment by jklymak** [comment]
Hmm, I thought opening closing was supposed to restart the tests, but it didn't restart the doc builds...  A little leery about merging w/o the tests...

**Comment by phobson** [comment]
please work 🤞

**Comment by QuLogic** [comment]
CircleCI doesn't test merge commits (which is a bit unfortunate), and that means if you close and re-open, it doesn't restart the test (since the head of the PR hasn't changed). You can restart in the UI, but it still won't see any new things in `master`.
